### ISSUE TITLE
FabricModelPredicateProviderRegistry only accepts ClampedItemPropertyFunctions

### ISSUE THREAD

**Issue by quat1024** [FabricMC/fabric#2107]
(Mojmap.)

Open up your editor and pull up the class `ItemProperties`. You will see:

* two `Map<ResourceLocation, ItemPropertyFunction>` storages, `GENERIC_PROPERTIES` and `PROPERTIES`
* two helper methods that register things to this storage, `registerGeneric` and `register`, that both take *Clamped*ItemPropertyFunction as a parameter.

(For context: ClampedItemPropertyFunction is a helper interface that extends ItemPropertyFunction, overriding its `call` method with one that delegates to a new `unclampedCall` method, then clamps the result to 0..1. It's important to note that ClampedItemPropertyFunction is a helper interface only. It's just an implementation detail. Everything in Minecraft that *uses* item property functions, and the backing storage for item property functions, uses the more generic ItemPropertyFunction interface instead.)

Anyway. Given this setup, there are two ways to design something like `FabricModelPredicateProviderRegistry`:

* Accept ItemPropertyFunctions, and write them directly into the backing storages (GENERIC_PROPERTIES and PROPERTIES)
* Accept *Clamped*ItemPropertyFunctions, and call the register functions that take clamped functions.

FabricModelPredicateProviderRegistry took the second approach, which limits its scope to `ClampedItemPropertyFunctions` only.

I want to register an item property function that outputs numbers greater than 1, so... this is a problem!

[The obvious workaround of manually overriding `ClampedItemPropertyFunction` with a `call` that delegates straight through without doing any clamping](https://github.com/quat1024/incorporeal-3/blob/8db27281f77a493a1ec3bb9f7956a0c5191cfe89/Fabric/src/main/java/agency/highlysuspect/incorporeal/platform/fabric/client/FabricClientBootstrapper.java#L30-L51) *works*, but if I wanted to spend my time fighting my modding API, I wouldn't be writing Fabric mods ;)

For comparison, [Forge's `ItemProperties` patch](https://github.com/MinecraftForge/MinecraftForge/blob/1.18.x/patches/minecraft/net/minecraft/client/renderer/item/ItemProperties.java.patch) simply changes the method signature of `register` to not take `Clamped` functions, and that's pretty much all it has to do. It would be easy Fabric-side to accessor and write to the backing storages directly instead of calling the helper functions.

**Comment by Technici4n** [comment]
`ItemPropertyFunction` is deprecated in vanilla, so not sure if we should be using it. 🤔

**Comment by quat1024** [comment]
Yes, but it's also used (compasses, pretty sure?) so i think its just mojang being weird with `@Deprecated` again

**Comment by PepperCode1** [comment]
Item model properties were completely reworked along with the rest of the item model system in Minecraft 1.21.4; none of the classes referenced by this issue exist anymore. There is also no similar issue with any new APIs as far as I can see.
